Frameworks/AddressBook/ABAddressBookManagerInternal.mm

```diff
@@ -82,4 +189,171 @@ - (NSArray*)getListOfContacts {
+
+            _setError(error, @"Only existing contacts can be removed from the AddressBook.\n");
+            return false;
+        }
+        case kAddressBookReadWriteContact: {
+            __ABContactOperation* operation = _operations[contact.contact.id];
+
+            // A read/write contact can be deleted in 2 cases: if no operation was associated,
+            // or if the associated operation is that it has been modified. In the case that
+            // it is already marked to delete, deleting twice doesn't make sense.
+            if (operation == nil) {
+                operation = [[__ABContactOperation alloc] initWithContact:contact.contact shouldDelete:YES];
+                _operations[contact.contact.id] = operation;
+                return true;
+            } else if (operation.shouldDelete) {
+                _setError(error, @"This contact has already been deleted.\n");
+                return false;
+            } else {
+                operation.shouldDelete = YES;
+                return true;
+            }
+        }
+        case kAddressBookReadOnlyContact:
+        default:
+            _setError(error,
+                      @"Read-only contacts cannot be deleted. Use ABAddressBookCopyArrayOfAllUserAppPeople to get contacts that can "
+                      @"be deleted.\n");
+            return false;
+    }
+}
+
+- (void)modifyContact:(ABRecordRef)record {
+    _ABContact* contact = (__bridge _ABContact*)record;
+    __ABContactOperation* operation = _operations[contact.contact.id];
+
+    // If there was no operation previously associated with this record, then
+    // set it to be modified. The other two cases are that the contact was
+    // already to be modified (in which case, we don't need to do anything),
+    // or that the contact is marked to be deleted -- in this case, since
+    // it is to be deleted, there is no need to worry about any modifications.
+    if (operation == nil) {
+        operation = [[__ABContactOperation alloc] initWithContact:contact.contact shouldDelete:NO];
+        _operations[contact.contact.id] = operation;
+    }
+}
+
+- (bool)hasUnsavedChanges {
+    return _toAdd.count + _operations.count > 0;
+}
+
+- (void)revert {
+    [_toAdd removeAllObjects];
+    [_operations removeAllObjects];
+}
+
+- (bool)save {
+    // TODO #942: When the projections for WACContactList's saveContactAsync and deleteContactAsync
+    // get success/fail blocks, use a dispatch_group_t to spawn off all async calls at the same
+    // time rather than the blocking WFIASyncAction getResults, and return false in the case
+    // that the fail block is called (rather than just returning true always).
+
+    // Go through all contacts that were newly created and need to be added,
+    // and save them to this app's contact list.
+    for (WACContact* contact in _toAdd) {
+        RTObject<WFIAsyncAction>* asyncAction = [_contactList saveContactAsync:contact];
+        [asyncAction getResults];
```

Comment:
Oh, I see. TODO above.

---
In reply to: [77100271](https://github.com/Microsoft/WinObjC/pull/935#discussion_r77100271) [](ancestors = 77100271)